### PR TITLE
Apigw dynamodb cache invalidation

### PR DESCRIPTION
**Pull request by RedHeadDigital**
*Issue #, if available:*

*Description of changes:*

Introduce new Serverless pattern for function-less API Gateway cache invalidation pattern using DynamoDB Streams, EventBridge Pipes & EventBridge.

apigw-cache-invalidation-eventbridge-architecture
Here is an architecture diagram of the solution, including technical references and rough data flow.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by tmmruk** [comment]
Hi Mark, 
As requested, here is my feedback on this pattern:

- Add the architecture diagram to the README file as it clearly explains the solution. The diagram added to serverlessland will be a simplified version of this so it's good to include the fully annotated version in the README.
- Be explicit about the hourly charge incurred for the API Gateway cache and resources should be cleaned up at the end to avoid costs.
- The 'hello world' POST request should be send to the ApiUrl/**pets** - make sure to state this in the README.
- Use 'sam delete' to cleanup the stack as this will also delete the package uploaded to S3.

Otherwise, looks good to me and all working in my tests! Thanks for contributing this code to ServerlessLand!

Tom

File: apigw-cache-invalidation-eventbridge/README.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+> **Note:** This process has been implemented with the assumption that the update record can be published into a wider architecture and consumed by many consumers through use of the event bus and suitable filters.
+
+![Detailed architecture diagram](img/apigw-cache-invalidation-eventbridge-architecture.png)
+
+## Testing
+
+1. After the application is deployed, grab the ApiUrl endpoint from the outputs. If you missed it, simple use the SAM list command to retrieve.
+
+```
+sam list outputs
```
Comment (parikhudit) [inline apigw-cache-invalidation-eventbridge/README.md:61]: Did you mean ```sam list stack-outputs --stack-name <stack_name>```?

File: apigw-cache-invalidation-eventbridge/pattern.json [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  "testing": {
+    "text": ["See the GitHub repo for detailed testing instructions."]
+  },
+  "cleanup": {
+    "text": ["sam delete --stack-name STACK_NAME"]
+  },
+  "authors": [
+    {
+      "name": "Mark White",
+      "image": "https://media.licdn.com/dms/image/C4D03AQE17W81CnvT0w/profile-displayphoto-shrink_400_400/0/1633262092892?e=1706140800&v=beta&t=m6z_Nel7szyg1sPhngIvhkGE9cnpgU65WuaVn6Eo110",
```
Comment (parikhudit) [inline apigw-cache-invalidation-eventbridge/pattern.json:48]: Could you share an alternate URL? This URL isn't accessible.

File: apigw-cache-invalidation-eventbridge/pattern.json [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      "bio": "Principal Engineer at Dunelm",
+      "linkedin": "https://www.linkedin.com/in/mark--white/"
```
Comment (parikhudit) [inline apigw-cache-invalidation-eventbridge/pattern.json:50]: Please update this to only contain linkedIn id e.g. "mark--white" and not URL

File: apigw-cache-invalidation-eventbridge/pattern.json [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      "Once the event is available on the custom EventBridge, a Rule can consume the event and target the API Gateway with a `Cache-Control: max-age=0` header and correct AWS IAM permissions `execute-api:InvalidateCache` using the Arn method for the Target.",
+      "Note: This application uses resources that are not eligible for the AWS Free Tier"
+    ]
+  },
+  "gitHub": {
+    "template": {
+      "repoURL": "https://github.com/aws-samples/serverless-patterns/tree/main/apigw-cache-invalidation-eventbridge",
+      "templateURL": "serverless-patterns/apigw-cache-invalidation-eventbridge",
+      "projectFolder": "apigw-cache-invalidation-eventbridge",
+      "templateFile": ""
```
Comment (parikhudit) [inline apigw-cache-invalidation-eventbridge/pattern.json:21]: Add template.yaml?

**Comment by parikhudit** [comment]
@RedHeadDigital Thanks for the contribution and a good pattern overall. Could you review few minor comments I've added and update the PR? Thanks.
CC: @tmmruk

File: apigw-cache-invalidation-eventbridge/template.yaml [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+                    "properties": {
+                        "metrics": [
+                            [ "AWS/ApiGateway", "CacheMissCount", "ApiName", "DynamoDB Crud API", { "region": "${AWS::Region}" } ],
+                            [ ".", "CacheHitCount", ".", ".", { "region": "${AWS::Region}" } ]
+                        ],
+                        "view": "timeSeries",
+                        "stacked": false,
+                        "region": "${AWS::Region}",
+                        "period": 30,
+                        "stat": "Average",
```
Comment (parikhudit) [inline apigw-cache-invalidation-eventbridge/template.yaml:201]: Suggestion: Could you use stat as Sum for Cache Hit & Miss Count? Sum would provide actual value of requests that Hit or Missed the cache.

**Comment by RedHeadDigital** [comment]
Hi @parikhudit,
Thanks for the review and comments. I've addressed all the comments. 

Thanks
Mark

**Comment by mavi888** [comment]
Thanks for submitting your pattern.
Now this pattern is public in Serverless land

https://serverlessland.com/patterns/apigw-cache-invalidation-eventbridge


You can share your pattern using this link :)

**Pull request merged**